File: Companies/VMware/FindDuplicateFileinSystem.java
package Companies.VMware;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FindDuplicateFileinSystem {
    public List<List<String>> findDuplicate(String[] paths) {
        Map<String, List<String>> m = new HashMap<>();
        for (String path : paths) {
            String[] files = path.split(" ");
            for (int i = 1; i < files.length; i++) {
                String[] name_content = files[i].split("\\(");
                String name = name_content[0];
                String content = name_content[1].replace(")", "");
                List<String> l = m.getOrDefault(content, new ArrayList<>());
                l.add(files[0] + "/" + name);
                m.put(content, l);
            }
        }
        List<List<String>> re = new ArrayList<>();
        for (String k : m.keySet()) {
            if (m.get(k).size() > 1) {
                re.add(m.get(k));
            }
        }
        return re;
    }
}

File: Companies/VMware/FindDuplicateFileinSystem.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <sstream>

using namespace std;

class FindDuplicateFileinSystem
{
public:
  vector<vector<string>> findDuplicate(vector<string> &paths)
  {
    unordered_map<string, vector<string>> files{};
    for (auto path : paths)
    {
      stringstream ps{path};
      string root{}, s{};
      getline(ps, root, ' ');
      while (getline(ps, s, ' '))
      {
        string name = root + '/' + s.substr(0, s.find('('));
        string content = s.substr(s.find('(') + 1, s.find(')') - s.find('(') - 1);
        files[content].push_back(name);
      }
    }
    vector<vector<string>> re{};
    for (auto file : files)
    {
      if (file.second.size() > 1)
      {
        re.push_back(file.second);
      }
    }
    return re;
  }
};